### PR TITLE
loadsave code project

### PR DESCRIPTION
**Pull request by abuts**
The purpose of the job is to avoid writing the same serialization code for all Horace/Herbert classes.

The PR is not a complete job but the proposal to be expanded on all instrument classes and custom classes.  Please note how sample IX_aperture class lost 2/3 of its code.
So I am inviting the participants to look at the changes, discuss and use in their development. 

I will also use these methods for converting classes into upcoming `nxsqw` file format.

### CODE REVIEW

File: _test/test_instrument_classes/test_IX_aperture.m [hunk 1]
```diff
@@ -45,6 +69,59 @@ function test_pdf (self)
             assertEqualToTol(stdev.^2, [0.1^2;0.06^2]/12, 'reltol', 1e-3);
         end
         
+        function test_apperture_array_prev_versions(obj)
```
Comment (cmarooney-stfc) [inline _test/test_instrument_classes/test_IX_aperture.m:72]: aperture 1P just to keep it clean

**Review by cmarooney-stfc — CHANGES_REQUESTED** [review]
Presuming you will do the rest of the instrument bits (or leave it to me) once this infrastructure is working

File: _test/test_instrument_classes/check_matfile_IO.m [hunk 2]
```diff
@@ -0,0 +1,33 @@
+function [ok,mess] = check_matfile_IO (ver_str, save_variables, filepath,varargin)
+% Save to or read from mat file
+% input
+% ver_str        -- the name of the version string e.g. ver0 for old
```
[the remark this answers is not in the document]
Comment (abuts) [reply _test/test_instrument_classes/check_matfile_IO.m:4]: that's Toby's function, extracted from one of the tests, disassembled in pieces. But its probably correct to write better description to its arguments

File: _test/test_instrument_classes/check_matfile_IO.m [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+%                   version ver1 for version 1 etc
+% save_variables -- if true, saves sample variables, if false load
+%                   them from disk
```
[the remark this answers is not in the document]
Comment (abuts) [reply _test/test_instrument_classes/check_matfile_IO.m:7]: its save/load here as save/load operators are tested

File: _test/test_instrument_classes/check_matfile_IO.m [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+% filepath       -- the location of source test files on hdd.
+%                   If the routine writes to disk, it writes
+%                   test files into temporarty directory
+% varargin       -- variables to save/load on hdd
+mess = '';
+ok = true;
+for i=1:numel(varargin)
+    class_name = class(varargin{i});
+    arg_name = inputname(i+3);
```
[the remark this answers is not in the document]
Comment (abuts) [reply _test/test_instrument_classes/check_matfile_IO.m:16]: that's Matlab intrinsic function

File: _test/test_instrument_classes/check_matfile_IO.m [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    flname = [ver_str,'_',class_name,'_',arg_name,'.mat'];
+    if save_variables
+        eval([arg_name,' = varargin{i};']);
+        try
+            save(fullfile(tmp_dir,flname),arg_name);
```
[the remark this answers is not in the document]
Comment (abuts) [reply _test/test_instrument_classes/check_matfile_IO.m:21]: no, its the specific function, used by these particular tests. You do not want to give it options -- will be too complicated&unnecessary

File: herbert_core/utilities/classes/@serializable/serializable.m [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+            %          recover from input S
+            %
+            % Output:
+            % -------
+            %   obj     Either (1) the object passed without change, or (2) an
+            %           object (or object array) created from the input structure
+            %       	or structure array)
+            if isobject(S) && isa(S,class(class_instance))
+                obj = S;
+            else % call private implementation
```
[the remark this answers is not in the document]
Comment (abuts) [reply herbert_core/utilities/classes/@serializable/serializable.m:122]: ? do not understand this remark. The `else` is calling private function `loadobj_` which does the rest of the processing.  For what it does look at `loadobj_` one does not want to repeat the same description in two independent places.

**Pull request merged**